### ISSUE TITLE
Abstract Tooltip HTML (and any other HTML) into external template files

### ISSUE THREAD

**Issue by joshlangner** [c3js/c3#697]
We should abstract Tooltip and any other HTML into external templates which can be referenced from the source. A global config param should be created which allows developers to point to different templates (e.g. their own HTML).

**Comment by morganpolotan** [comment]
+1

**Comment by aendrew** [comment]
I'm fairly in support of this, particularly if all the default templates are precompiled and bundled with the concatenated/minified `.js` files.

### Pros
+ Better access to C3 component styling and layout
+ Less DOM manipulation in the code
+ Probably easier maintenance long-term

### Cons
+ Possible performance implications (not sure TBH; a precompiled template might be faster than all the `.appends` throughout the library)
+ Severe possibility of introducing new bugs and breaking stuff
+ We'd have to decide on a templating language

My personal preference for templating library is Handlebars, namely because it's familiar to me and allows [precompilation](http://handlebarsjs.com/precompilation.html) (important for preventing users from having to include a bunch of HTML and stuff, which would lead to endless unnecessary issues in the queue). Granted, I haven't looked at templating languages in ages and have no idea what's current and good.

**Comment by joshlangner** [comment]
@aendrew 

I don't think we need to open the templating language debate, and we shouldn't need to add any additional bloat (e.g. templating engines). The simple goal should be to abstract the HTML / markup into its own file(s), which the user could then override with their own templating as they see fit.

The issue today is that the html is hardcoded into the source and in order to override, the user must find it in the source and copy/paste/override in a fairly complex way.

**Comment by aendrew** [comment]
@joshlangner Fair enough, but how would this get included with the library? Would the library have to load a lot of partials on page load or something? I can't help but feel there would be some performance implications until precompiled templates are included with the lib somehow, with those getting overridden if a template path is provided.
